### PR TITLE
[docs] Move page api version and metadata context to separate providers

### PR DESCRIPTION
**Pull request by byCedric**
# Why

Part of [ENG-3964](https://linear.app/expo/issue/ENG-3964/split-up-documentationpagetsx)

The `DocumentationPage` does a lot of things, like tracking current selected API version, scroll events, layout, etc. This PR moves the API version code to its own provider.

# How

- Added `providers/page-api-version`
- Moved `components/DocumentationPageContext` to `providers/page-metadata`
- Implemented version context in `AlgoliaSearch`, `VersionSelector`, and `DocumentationSidebar`
- Updated page metadata implementations
- Removed existing version code from `DocumentationPage` and siblings

# Test Plan

## New PageApiVersion context provider
- Open a non-versioned API page by URL → _version should default to `latest`_
- Open an API reference page by clicking → _latest should be rendered in sidebar and selected_
- Change API version with the VersionSelector → _version should change + URL change + page reload with the right version_
- Open any API reference page with different version → _version should be infered from URL_

## New PageMetadata context provider
- Open a page with API section or InstallSection components → _packageName should be inferred from context_

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: docs/components/DocumentationPage.tsx [hunk 1]
```diff
@@ -49,24 +49,26 @@ const HIDDEN_ON_DESKTOP = css`
   }
 `;
 
-type Props = {
+type Props = React.PropsWithChildren<{
```
Comment (byCedric) [inline docs/components/DocumentationPage.tsx:52]: So I guess we did this one wrong all along 🤷

File: docs/providers/api-version.tsx [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  version: 'latest',
+  setVersion: (newVersion: string) => console.error('ApiVersion context provider not found'),
+});
+
+export function ApiVersionProvider(props: Props) {
+  const router = useRouter();
+  const [version, setVersion] = useState('latest');
+
+  const onVersionChange = useCallback((newVersion: string) => {
+    if (isVersionedUrl(router.pathname)) {
```
Comment (byCedric) [inline docs/providers/api-version.tsx:25]: We could also check if the version is valid, but I like to think our code only uses valid versions

File: docs/providers/api-version.tsx [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+      {props.children}
+    </context.Provider>
+  );
+}
+
+export function useApiVersion() {
+  return useContext(context);
+}
+
+// TODO(cedric): remove the need for an HOC component
```
Comment (byCedric) [inline docs/providers/api-version.tsx:56]: 🥲

File: docs/common/utilities.ts [hunk 4]
```diff
@@ -43,7 +43,7 @@ export const replaceVersionInUrl = (url: string, replaceWith: string) => {
 };
 
 export const getVersionFromUrl = (url: string) => {
-  return url.split('/')[2];
+  return new URL(url, 'https://docs.expo.dev').pathname.split('/')[2];
```
Comment (byCedric) [inline docs/common/utilities.ts:46]: This didn't handle urls with query parameters, like `https://docs.expo.dev/versions/v44.0.0/sdk/notifications?redirected=`. 🤷

File: docs/providers/api-version.tsx (hunk 3 above)
Comment (byCedric) [reply docs/providers/api-version.tsx]: removed

File: docs/components/DocumentationPage.tsx [hunk 5]
```diff
@@ -50,23 +49,22 @@ const HIDDEN_ON_DESKTOP = css`
...
+  version: string;
+}>;
 
 type State = {
   isMenuActive: boolean;
   isMobileSearchActive: boolean;
 };
 
-export default class DocumentationPage extends React.Component<Props, State> {
+class DocumentationPageWithApiVersion extends React.Component<Props, State> {
```
Comment (jonsamp) [inline docs/components/DocumentationPage.tsx:67]: Any reason why we're leaving this as a class? Could make this all one function component with the API version hook. Totally okay if you were just preventing scope creep, just wondering if there was a reason.

File: docs/components/VersionSelector.tsx [hunk 6]
```diff
@@ -49,35 +50,42 @@ const STYLES_SELECT_ELEMENT = css`
...
-          {Utilities.getUserFacingVersionString(v, LATEST_VERSION, BETA_VERSION)}
-        </option>
-      ))}
-    </select>
-  </div>
-);
+const VersionSelector: React.FC<Props> = ({ style }) => {
+  const { version, hasVersion, setVersion } = usePageApiVersion();
+
+  // Don't render the version selector on non-versioned pages
```
Comment (jonsamp) [inline docs/components/VersionSelector.tsx:59]: I don't think this comment is necessary. the code is well named/written.

File: docs/components/VersionSelector.tsx [hunk 7]
```diff
@@ -49,35 +50,42 @@ const STYLES_SELECT_ELEMENT = css`
...
+
+  return (
+    <div css={STYLES_SELECT} style={style}>
+      <label css={STYLES_SELECT_TEXT} htmlFor="version-menu">
+        <div>{Utilities.getUserFacingVersionString(version, LATEST_VERSION, BETA_VERSION)}</div>
+        <ChevronDownIcon style={{ height: '16px', width: '16px' }} />
+      </label>
+      {
+        // hidden links to help test-links spidering
+        VERSIONS.map(v => (
```
Comment (jonsamp) [inline docs/components/VersionSelector.tsx:72]:
```suggestion
        VERSIONS.map(version => (
```
I much prefer to write out the variable name for readability

File: docs/components/VersionSelector.tsx [hunk 8]
```diff
@@ -49,35 +50,42 @@ const STYLES_SELECT_ELEMENT = css`
...
+          <a key={v} style={{ display: 'none' }} href={`/versions/${v}/`} />
```
Comment (jonsamp) [inline docs/components/VersionSelector.tsx:73]:
```suggestion
          <a key={v} style={{ display: 'none' }} href={`/versions/${version}/`} />
```

File: docs/components/VersionSelector.tsx [hunk 9]
```diff
@@ -49,35 +50,42 @@ const STYLES_SELECT_ELEMENT = css`
...
+  // Don't render the version selector on non-versioned pages
+  if (!hasVersion) {
+    return null;
+  }
+
+  return (
+    <div css={STYLES_SELECT} style={style}>
+      <label css={STYLES_SELECT_TEXT} htmlFor="version-menu">
+        <div>{Utilities.getUserFacingVersionString(version, LATEST_VERSION, BETA_VERSION)}</div>
+        <ChevronDownIcon style={{ height: '16px', width: '16px' }} />
```
Comment (jonsamp) [inline docs/components/VersionSelector.tsx:68]: Would love to take these styles out into named CSS styles.

Also, there's a `<ChevronDown />` icon exported from @expo/styleguide. This code would be turned into:

```tsx
import { ChevronDown, iconSize } from '@expo/styleguide';

// ...

<ChevronDown size={iconSize.small} />
```

File: docs/components/VersionSelector.tsx [hunk 10]
```diff
@@ -49,35 +50,42 @@ const STYLES_SELECT_ELEMENT = css`
...
+      </label>
+      {
+        // hidden links to help test-links spidering
```
Comment (jonsamp) [inline docs/components/VersionSelector.tsx:71]: what does this comment mean?

in general, I think we should use comments to explain one-off exceptional behavior (like a comment explaining that Safari renders something incorrectly, so adding a style to fix it). With that, I don't think we need comments in most places. If it's not clear what this code does, I wonder if we could add a variable name to help explain.

File: docs/package.json [hunk 11]
```diff
@@ -80,10 +81,11 @@
     "eslint-config-universe": "^10.0.0",
     "eslint-plugin-lodash": "^7.3.0",
     "http-server": "^0.12.3",
-    "jest": "26.1.0",
+    "jest": "^27.4.7",
```
Comment (jonsamp) [inline docs/package.json:84]: thanks for updating deps

File: docs/providers/page-api-version.test.tsx [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+      '/versions/v44.0.0/sdk/notifications'
+    );
+  });
+
+  it('returns new pathname for react-native pathname', () => {
+    expect(replaceVersionInPath('/versions/v44.0.0/react-native/stylesheet/', 'v43.0.0')).toBe(
+      '/versions/v43.0.0/react-native/stylesheet/'
+    );
+  });
+});
```
Comment (jonsamp) [inline docs/providers/page-api-version.test.tsx:131]: thanks for adding a test!

File: docs/providers/page-api-version.tsx [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+export function isVersionedPath(path: string) {
+  return path.startsWith('/versions/');
+}
+
+/**
+ * Find the version within the pathname of the URL.
+ * This only accepts pathnames, without hashes or query strings.
+ */
+export function getVersionFromPath(path: string) {
+  return !isVersionedPath(path) ? null : path.split('/', 3).pop()!;
```
Comment (jonsamp) [inline docs/providers/page-api-version.tsx:54]:
```suggestion
  const versionCaptureRegex = /\/versions\/v((\d+|\.)*)?/
  return !isVersionedPath(path) ? null : path.match(versionCaptureRegex)[1];
```

Just another way to write this that may be a little clearer. I'm hoping we could add a variable to make this more readable.

**Review by jonsamp — APPROVED** [review]
Thanks for adding context in instead of prop drilling data through the app. This is a great improvement. I left a lot of little comments. Feel free to take or leave them, or complete them in a different PR.

File: docs/components/DocumentationPage.tsx (hunk 5 above)
Comment (byCedric) [reply docs/components/DocumentationPage.tsx:67]: It still has a lot of old code to fully refactor this to a functional component. We still need to refactor:
- The scroll refs and the scroll layout
- Selecting the category of navigation by url
- Set header meta tags for a page
- Rendering mobile layout vs desktop
- Smooth scroll to a header by ID

That's what this component is handling, besides ofc, rendering everything 🙈

File: docs/components/VersionSelector.tsx (hunk 9 above)
Comment (byCedric) [reply docs/components/VersionSelector.tsx:68]: Absolutely! But to refactor the icons, we need them in the doc. It's currently in PR #16134, but that's blocked by both PR #16131 and PR #16118. So for now, I kept the original version selector but with the new context.

File: docs/components/VersionSelector.tsx (hunk 10 above)
Comment (byCedric) [reply docs/components/VersionSelector.tsx:71]: This is from the current version selector code. I didn't write it, but I believe this influenced SEO. By adding hidden links to all versions, the spider should easily pick up multiple versions. Without these hidden links, we don't have any direct URL reference to the other Expo SDK versions.

Definitely can improve that comment though.

File: docs/package.json (hunk 11 above)
Comment (byCedric) [reply docs/package.json:84]: I kinda wanted it to test the code 😬  It works pretty well, especially with the newer `jest.mocked` method!

File: docs/providers/page-api-version.tsx (hunk 13 above)
Comment (byCedric) [reply docs/providers/page-api-version.tsx:54]: I see what you mean, but the unfortunate thing here is that these values are all valid versions:
- `v44.0.0` -> captured well by this regex
- `v41.0.0` -> same, well captured
- `latest` -> not captured
- `unversioned` -> not captured

To create a regex out of that, I think this might be a bit easier 🙈  Definitely ok with swapping this for regexes, maybe a good follow-up PR?

File: docs/components/VersionSelector.tsx (hunk 6 above)
Comment (byCedric) [reply docs/components/VersionSelector.tsx:59]:
```suggestion
```

File: docs/components/VersionSelector.tsx (hunk 10 above)
Comment (byCedric) [reply docs/components/VersionSelector.tsx:71]:
```suggestion
        // Add hidden links to create crawler references to all other SDK versions
        // This should help search bots help find SDK versions, while using JS to switch between them

File: docs/components/VersionSelector.tsx [hunk 14]
```diff
@@ -49,35 +50,42 @@ const STYLES_SELECT_ELEMENT = css`
...
+      <select
+        id="version-menu"
+        css={STYLES_SELECT_ELEMENT}
+        value={version}
+        onChange={e => setVersion(e.target.value)}>
+        {VERSIONS.map(v => (
+          <option key={v} value={v}>
+            {Utilities.getUserFacingVersionString(v, LATEST_VERSION, BETA_VERSION)}
+          </option>
+        ))}
```
Comment (byCedric) [inline docs/components/VersionSelector.tsx:85]:
```suggestion
        {VERSIONS.map(version => (
          <option key={version} value={version}>
            {Utilities.getUserFacingVersionString(version, LATEST_VERSION, BETA_VERSION)}
          </option>
        ))}
```
